### DISCUSSION TITLE
New language feature suggestion: the `shapeof` keyword

### DISCUSSION THREAD

**Discussion by johnknoop** [dotnet/csharplang#5886] [Language Ideas]
When records were announced a few years ago surrounded by words such as "structural equality", my expectation was that we would be able to compare any two records as long as they had the same shape, like in structurally typed languages. This expectation was perhaps foolish (I have no idea how hard it would be to implement such a feature in C#). But the reality that records are little more than classes continue to bug me to this day, since I still have to spend a lot of time writing code that maps an object of one record type to another record type with the same shape.

## The painpoint

Here's a small example that is pretty common in C# codebases. I have a ASP.NET project sitting around a domain model, and the domain model is of course ignorant of the ASP.NET assembly that hosts it. I want to be able to change my domain model without breaking my public API, so therefore I have two separate sets of types involved in each operation. In this example the `CreateOrderCommand` which lives in the domain model, and then the `CreateOrderRequest` which lives in the webapp.

```cs
namespace ExampleDomainModel {
    public record TaxAmount(int TaxRate, decimal Amount);
    public record OrderLine(string ProductId, int Quantity, decimal Subtotal, decimal Total, IList<TaxAmount> Taxes);
    public record CreateOrderCommand(IList<OrderLine> Lines, decimal Subtotal, decimal Total, IList<TaxAmount> Taxes);

    public class MyAggregate
    {
        public void CreateOrder(CreateOrderCommand command) {
            // ...
        }
    }
}
```

```cs
namespace WebApi {
    public record TaxAmount(int TaxRate, decimal Amount);
    public record OrderLine(string ProductId, int Quantity, decimal Subtotal, decimal Total, IList<TaxAmount> Taxes);
    public record CreateOrderRequest(IList<OrderLine> Lines, decimal Subtotal, decimal Total, IList<TaxAmount> Taxes);

    public class MyController
    {
        [HttpPost]
        public async Task<IActionResult> CreateOrder(CreateOrderRequest request) {
            /*
            * I now need to do all this manual mapping, even though it's structurally the exact same information.
            * Sure, I could use something like AutoMapper, but then I don't get compile-time checking for mismatches
            */
            myAggregate.CreateOrder(new ExampleDomainModel.CreateOrderCommand(
                Lines: request.Lines.Select(line => new ExampleDomainModel.OrderLine(
                        ProductId: line.ProductId,
                        Quantity: line.Quantity,
                        Subtotal: line.Subtotal,
                        Total: line.Total,
                        Taxes: line.Taxes.Select(tax => ExampleDomainModel.TaxAmount(tax.TaxRate, tax.Amount)).ToList()
                    )).ToList(),
                Subtotal: request.Subtotal,
                Total: request.Total,
                Taxes: request.Taxes.Select(tax => ExampleDomainModel.TaxAmount(tax.TaxRate, tax.Amount)).ToList()
            ));
            /*
            * This is just a small example. In the real world, these mappings often grow quite large.
            */
        }
    }
}
```
Most of the code in the example above is just stupid mapping.

## The suggestion

I would like to suggest a new language feature: the `shapeof` keyword which you can add in front of any parameter type in a method signature, which tells the compiler to not care about the nominal type of the argument that gets passed in, but only the structure of it.

So with the `shapeof` keyword, the `CreateOrder` method in my domain model would look like this:

```cs
public void CreateOrder(shapeof CreateOrderCommand command) {
    // ...
}
```

Which allows me to reduce my WebAPI method down to this:

```cs
[HttpPost]
public async Task<IActionResult> CreateOrder(CreateOrderRequest request) {
    myAggregate.CreateOrder(request);
}
```

If there were some structural difference between `ExampleDomainModel.CreateOrderCommand` and `WebApi.CreateOrderRequest`, like a slithgly different naming or a `double` instead of a `decimal` somewhere, then I'd get a compilation error, which I could solve using the `with` expression:

```cs
[HttpPost]
public async Task<IActionResult> CreateOrder(CreateOrderRequest request) {
    myAggregate.CreateOrder(request with {
        MyProperty = request.MyDifferentlyNamedProperty
    });
}
```

I understand that adding a feature like this to a nominally typed language is a challenge. Maybe a compilation step that autogenerates an `interface ICreateOrderCommandAutoGenerated` and a `class CreateOrderRequestAutoGenerated` that implements it? This of course needs to be completely sourcemapped so that these autogenerated types aren't exposed to the developer in debug-time.

**Comment by Xyncgas** [comment]
Maybe in the future we can write ```I am in Love with the shapeof (you)```

**Comment by TahirAhmadov** [comment]
This is a question which has been asked for years now, since the first ASP.NET MVC. It looks like while the domain model (AKA business layer) is ignorant of the ASP.NET app (AKA presentation layer), as it should be, the presentation layer is aware of the business layer. Why not use the records from the business layer in the presentation layer in these situations of perfect one-to-one match? In my projects, I use a combination of 80% entity classes (which represent the actual database tables), 18% business layer DTOs (your typical `OrderRequest`), and 2% special presentation layer models (things that are tied to presentation only). The main benefit of my approach is the reduction of large amount of copy-pasta (DRY principle) and avoiding the need for "object mappers/converters".
I would not support a language feature to accomplish what you are trying to do. You should consider an object mapping library if you absolutely must redefine all your classes.

**Reply by johnknoop** [reply]
I think we should try and keep this thread about language design, but in short: if you have a public API that you don't wanna break, while at the same time you want to be able to make changes to your domain model, then you don't want them coupled like that.

There a a lot of workarounds (like AutoMapper as mentioned above) but they all come with drawbacks. Language support for structural typing would be the ultimate solution.

**Reply by TahirAhmadov** [reply]
If you have to make a change to your domain model which includes modifying that `OrderRequest` DTO, you will **have** to also change the API. How can you add or remove a property, for example, without exposing that change to the API?
Imagine all the drawbacks of AutoMapper and now imagine how difficult it would be to make it work as part of the language. Language has to work consistently, with no unexpected edge cases, etc. With AutoMapper, you can configure "conventions", etc. None of that is possible with the language proposal. In fact, I think adding this to the language would be a terrible solution.

**Reply by johnknoop** [reply]
The CreateOrderRequest isn't part of the domain model. That's the whole point. But like I said, I'd like this discussion to be about language design and not suggestions of workarounds. But thanks for your input.

**Reply by HaloFour** [reply]
The discussion of language design generally includes existing solutions, even if they are considered "workarounds".  Language changes are very expensive, if there are solutions that fit the need without requiring changing the language spec and compiler they will be taken into consideration.

**Reply by johnknoop** [reply]
Sure but if I was looking for help in that department I would post a question on Stack Overflow.

**Comment by HaloFour** [comment]
IMO I think this would fall under the various "shapes" proposals.  You'd define a "shape" of the schema, likely through an interface, and then you can specify that both of the existing types satisfy that shape.  Via roles/witness you could also define extensions to map some of the members in a custom manner or fill in members that don't exist in the source type.

Here's an older discussion on shapes via extensions and roles:  https://github.com/dotnet/csharplang/discussions/5496

It's involved and the mentioned use cases don't quite align with this one but I think the same solution would fit here.

**Reply by johnknoop** [reply]
Not sure I agree with that. This suggestion isn't about defining shapes in any other ways than what's currently possible.

**Reply by HaloFour** [reply]
Any proposal that touches on structural equality is about defining shapes.  You want two disparate types to fit a specified structure, regardless of where that comes from, and the ability to override the mappings between the existing types and that structure, that is exactly what the linked discussion is looking to accomplish.

**Reply by johnknoop** [reply]
I still believe my suggestion is different enough from #5496 not to get it mixed up with that. For example: if I had to specify shapes using a separate syntax, then that would take me the same amount of time as I now have to spend on mapping from one object graph to another.

**Reply by HaloFour** [reply]
It is different, but I still think it's firmly under the same umbrella.  Since that discussion is still pretty early in design it's more likely that they'll consider this use case in that context vs. considering it completely separately.  It's also highly unlikely that the team would consider and implement two completely different features around structural typing.
